File: Not bad becomes good/not_bad_becomes_good.js
/* Given a string, find the first appearance of the substring 'not' and 'bad'. If the 'bad' follows the 'not', replace the whole 'not'...'bad' substring with 'good'. Return the resulting string. So 'This dinner is not that bad!' yields: This dinner is good!
*/

function replaceNotBadByGood(myString){
    // check if "not" exists
    var indexOfNot = myString.indexOf('not');    
    if (indexOfNot == -1){
        return myString;
    } 
    
    // check if "bad" exists after not
    var stringFromNot = myString.substring(indexOfNot, string.length +1);
    var indexOfBadInSubstring = stringFromNot.indexOf("bad");
    if (indexOfBadInSubstring == -1){
        return myString;
    } 
    var indexOfBad = myString.indexOf("bad");
    
    // replace "not ... bad" by "good"
    var stringToBeReplaced = myString.substring(indexOfNot, indexOfBad + "bad".length);
    return myString.replace(stringToBeReplaced, "good");
}

var testString = [
        "This dinner is not that bad!", 
        "This dinner is bad not that !",
        "Hello John!"
    ];
    
testString.forEach(function(str){
    console.log(str + " > " + replaceNotBadByGood(str));
});

File: not_bad_becomes_good/not_bad_becomes_good.py
def replaceString(myString):
    message = myString;
    
    try:
        indexOfNot = myString.index('not');
        
        try:
            indexOfBad = myString[indexOfNot:].index('bad');
            print indexOfBad;
            message += " > " + myString.replace(myString[indexOfNot:indexOfNot + indexOfBad + 3], "good");            
        except ValueError:
            message += " > " + "Sorry, but we can't find bad after not in this sentence";        
        
    except ValueError:
        message += " > " + "Sorry, but we can't find not in this sentence";
        
    print message + "\n";


myStringOK = "This dinner is not that bad";
myStringNotOK1 = "This dinner is bad not that ";
myStringNotOK2 = "Hello world!";
replaceString(myStringOK);
replaceString(myStringNotOK1);
replaceString(myStringNotOK2);